src/wayland/shell/xdg/mod.rs

```diff
@@ -615,6 +618,28 @@ impl ToplevelStateSet {
             true
         }
     }
+
+    /// Filter the states according to the provided version
+    /// of the [`XdgToplevel`]
+    pub(crate) fn into_filtered_states(self, version: u32) -> Vec<xdg_toplevel::State> {
+        let is_tiled = |state: &xdg_toplevel::State| {
+            matches!(
+                state,
+                xdg_toplevel::State::TiledTop
+                    | xdg_toplevel::State::TiledBottom
+                    | xdg_toplevel::State::TiledLeft
+                    | xdg_toplevel::State::TiledRight
+            )
+        };
+
+        let contains_tiled = self.states.iter().any(|state| is_tiled(state));
+
+        if version >= XDG_TOPLEVEL_STATE_TILED_SINCE || !contains_tiled {
+            return self.states;
+        }
+
+        self.states.into_iter().filter(|state| !is_tiled(state)).collect()
```

Comment:
This is the kind of situation where we'd love `drain_filter` to be stable...

However, I think we could optimize for the happy path of the client using a recent xdg-shell version, in that case we can directly return `self.states` without even iterating the states once.

Reply:
Yeah, not the first time drain_filter would come handy 😩
I will split the check into two and move the happy path to the top 👍

Reply:
done